### PR TITLE
gh-126119: fix some crashes in code objects if `co_stacksize` is absurdly large

### PR DESCRIPTION
**Pull request by picnixz**
* Issue: gh-126119

### CODE REVIEW

File: Objects/frameobject.c [hunk 1]
```diff
@@ -1804,11 +1804,22 @@ PyDoc_STRVAR(clear__doc__,
...
+        // This could happen if the underlying code object has a
+        // very large stacksize (but at most INT_MAX) yet that the
+        // above offsets make the result overflow.
+        //
+        // This should normally only happen if PY_SSIZE_T_MAX == INT_MAX,
+        // but we anyway raise an exception on other systems for safety.
+        PyErr_SetString(PyExc_OverflowError, "size exceeds PY_SSIZE_T_MAX");
+        return NULL;
+    }
+    return PyLong_FromSsize_t(res + nslots * sizeof(PyObject *));
```
Comment (rruuaanng) [inline Objects/frameobject.c]: Are you sure you are using sizeof for **`PyObject *`**, not **`PyObject`**?
Use `sizeof()` for pointer types, which are always 4 bytes (32) or 8 bytes (64).

Comment (picnixz) [reply Objects/frameobject.c]: Yes, because the structure is only keeping pointers. If you were storing `PyObject` values directly, then it would be correct.

**Comment by picnixz** [comment]
I found out that this won't be sufficient so converting it to a draft for now until I patch more attackable entrypoints.

File: Objects/frameobject.c [hunk 2]
```diff
@@ -1804,11 +1804,22 @@ PyDoc_STRVAR(clear__doc__,
...
-    Py_ssize_t res;
-    res = offsetof(PyFrameObject, _f_frame_data) + offsetof(_PyInterpreterFrame, localsplus);
+    Py_ssize_t res = offsetof(PyFrameObject, _f_frame_data)
+                     + offsetof(_PyInterpreterFrame, localsplus);
     PyCodeObject *code = _PyFrame_GetCode(f->f_frame);
-    res += _PyFrame_NumSlotsForCodeObject(code) * sizeof(PyObject *);
-    return PyLong_FromSsize_t(res);
+    int nslots = _PyFrame_NumSlotsForCodeObject(code);
+    assert(nslots >= 0);
+    if ((size_t)nslots >= (PY_SSIZE_T_MAX - res) / sizeof(PyObject *)) {
```
Comment (ZeroIntensity) [inline Objects/frameobject.c]: I think this is UB. Signed integer overflow is undefined.

Comment (picnixz) [reply Objects/frameobject.c]: What would be UB? nslots is clipped to an int either way but we need to ensure that `nslots * sizeof(PyObject *)` does not overflow an *int* and not a `Py_ssize_t` (that's the issue IMO). 

The reason is due to the `assert(size < INT_MAX/sizeof(PyObject *));` in `_PyThreadState_PushFrame`.

Summary:

- `co_stacksize` is an int and thus is bounded by INT_MAX
- `co_framesize` is a function of `co_stacksize` but still < INT_MAX
-  `nslots` < `co_framesize` but we don't `sizeof(PyObject *) * nslots + res` to exceed a ssize_t

IIRC, POSIX doesn't specify that `sizeof(int) <= sizeof(ssize_t)` (nor does it guarantee `>=`). Since I don't really want to change the assertion on the size because otherwise this could mess up other parts of the code, we should make sure that the sizeof won't overflow.

Well.. technically, my fix is just about `con->stacksize >= (INT_MAX - nlocalsplus - FRAME_SPECIALS_SIZE) / sizeof(PyObject *)` but it didn't work for generator's frames actually.

Comment (ZeroIntensity) [reply Objects/frameobject.c]: If I'm understanding this right, we probably need to do something a little more complicated. I don't think it's even *possible* to check if something exceeds the `ssize_t` limit with > or <, because again signed integer overflow is undefined behavior. We probably need to cast things to uint64 and figure it out that way.

**Comment by markshannon** [comment]
How about rejecting excessively large `co_stacksize` when creating code objects?
It will be a much simpler fix.

**Comment by picnixz** [comment]
> How about rejecting excessively large co_stacksize when creating code objects?

That's what I'm actually doing (or am I not?)

File: Objects/frameobject.c (hunk 2 above)
Comment (picnixz) [reply Objects/frameobject.c]: Actually, I changed that one to an assertion.

**Comment by picnixz** [comment]
Ok, I found different issues. It's possible to bypass the check in the code object validator that I added when attacking generator's frames. I'll need a bit more time for that but I won't be coding anymore today.

The evil size that should be checked in code constructor is "((_testcapi.INT_MAX - 10) // sizeof(PyObject *)) - 1" but the following crashes:

```py
        ps = ctypes.sizeof(ctypes.c_void_p)  # sizeof(PyObject *)
        smallest_evil_co_stacksize = ((_testcapi.INT_MAX - 10) // ps)
        evil = f().gi_frame.f_code.__replace__(co_stacksize=smallest_evil_co_stacksize - 1) # ok
        evil_gi = types.FunctionType(evil, {})()  # crashes
        self.assertRaises(OverflowError, evil_gi.__sizeof__)  # crashes
``` 

With `smallest_evil_co_stacksize`, `__replace__` would raise as expected but with `smallest_evil_co_stacksize - 1` it's FunctionType that is crashing.

File: Objects/frameobject.c (hunk 2 above)
Comment (picnixz) [reply Objects/frameobject.c]: We are not checking the `ssize_t` limit actually. The assertion I want to avoid is `size < INT_MAX/sizeof(PyObject *)` not `size < PY_SSIZE_T_MAX/sizeof(PyObject *)`. 

You can *in advance* determine whether `X * Y` would overflow if you know both `X` and `Y` and the upper limit you want to impose on the output.

I don't see where you can have an undefined behaviour if you know that your quantities are all positive.

**Comment by picnixz** [comment]
The free-threaded build crashes but I don't know why... @colesbury do you have some idea? The bad code is 

```py
import types
def f(): yield
code = f().gi_frame.f_code
evil_code = code.__replace__(co_stacksize=268435444)  # OK
evil_gi_func = types.FunctionType(evil_code, {})      # OK
evil_gi = evil_gi_func()  							  # not ok!	
```

**Comment by ZeroIntensity** [comment]
I'll take a look at the feee-threaded failures.

**Comment by ZeroIntensity** [comment]
Ah @picnixz, the problem is [here](https://github.com/python/cpython/blob/8c22eba877225ab29cd64672dcb97f09a5f7336f/Include/internal/pycore_frame.h#L178). We need to overflow check, or perhaps prevent the code object from being able to do that in the first place.

**Comment by picnixz** [comment]
> We need to overflow check, or perhaps prevent the code object from being able to do that in the first place.

That's what I'm actually doing. It's exactly the check I'm performing but I actually why `_PyCode_Validate` is not called at this point! It seems that there is a validation that is not being done. Because `code->co_nlocalsplus + code->co_stacksize` is < `co_framesize` and the check is one `co_framesize`. I'll try to dive more into it.

**Comment by picnixz** [comment]
For the evil example above, I have `code->co_nlocalsplus + code->co_stacksize = 268435454` (which is `co_stacksize + FRAME_SPECIALS_SIZE` with `FRAME_SPECIALS_SIZE = 10`). Btw, the crash happens at `i = 4595`. Isn't it close to some recursion limit or stack limit or something like that?

File: Lib/test/test_frame.py [hunk 3]
```diff
@@ -1,5 +1,6 @@
 import copy
 import operator
+import platform
```
Comment (picnixz) [inline Lib/test/test_frame.py]:
```suggestion
```

File: Misc/NEWS.d/next/Core_and_Builtins/2024-10-29-11-47-19.gh-issue-126119.xbAvxt.rst [hunk 4]
```diff
@@ -0,0 +1,4 @@
+Fix a crash in DEBUG builds due to an overflow when the :attr:`co_stacksize
+<codeobject.co_stacksize>` field of a :ref:`code object <code-objects>` is
+set to an absurdely large integer.
```
Comment (picnixz) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-10-29-11-47-19.gh-issue-126119.xbAvxt.rst]:
```suggestion
set to an absurdly large integer.
```

File: Objects/codeobject.c [hunk 5]
```diff
@@ -436,7 +436,21 @@ _PyCode_Validate(struct _PyCodeConstructor *con)
...
+     * as framesize * sizeof(PyObject *) and assumed to be < INT_MAX. Thus,
+     * we need to dynamically limit the value of stacksize. Note that this
+     * usually prevents crashes due to assertions but a MemoryError may still
+     * be triggered later.
+     *
+     * See https://github.com/python/cpython/issues/126119 for details.
+     */
+    int ub = (int)(INT_MAX / sizeof(PyObject *)) - FRAME_SPECIALS_SIZE;
+    Py_ssize_t nlocalsplus = PyTuple_GET_SIZE(con->localsplusnames);
+    if (nlocalsplus >= (Py_ssize_t)ub || con->stacksize >= (int)ub - nlocalsplus) {
```
Comment (markshannon) [inline Objects/codeobject.c]:
```suggestion
    if (nlocalsplus + con->stacksize >= INT_MAX/2/sizeof(PyObject *)) {
```

File: Objects/codeobject.c [hunk 6]
```diff
@@ -436,7 +436,21 @@ _PyCode_Validate(struct _PyCodeConstructor *con)
...
+        PyErr_SetString(PyExc_OverflowError, "code: co_stacksize is too large");
```
Comment (markshannon) [inline Objects/codeobject.c]:
```suggestion
        PyErr_SetString(PyExc_OverflowError, "code: locals + stack size is too large");
```

**Review by markshannon — CHANGES_REQUESTED** [review]
We use `int` quite a lot in the compiler and VM, so beware 32 bit overflow.
We already assert that the size of the frame is less than `INT_MAX`: https://github.com/python/cpython/blob/main/Python/pystate.c#L2917
I can trigger that assert on main by increasing `co_stacksize` by one in https://github.com/python/cpython/pull/126122#issuecomment-2449623315

Checking that the combined locals + stacksize < `INT_MAX/16` should do.

We could increase the limit to something like `INT_MAX/8 - K` where `K` is roughly the base frame size, but I don't see any advantage to doing so and it would be more fragile.

Note: We should use `INT_MAX/16` rather than `INT_MAX/2/sizeof(PyObject *)`, otherwise it might be possible to generate a .pyc file on a 32 bit machine that wouldn't run on a 64 bit machine.

File: Objects/codeobject.c (hunk 5 above)
Comment (picnixz) [reply Objects/codeobject.c]: Won't I have an overflow if I have a lots of locals which would make `+` overflow? (well... the number of locals would be astronomically huge and I don't know if it's even possible to change them without declaring them). However, I can do

```C
    Py_ssize_t limit = (Py_ssize_t)(INT_MAX / 16) - FRAME_SPECIALS_SIZE;
    Py_ssize_t nlocalsplus = PyTuple_GET_SIZE(con->localsplusnames);
    if (nlocalsplus >= limit || con->stacksize >= limit - nlocalsplus) {
        PyErr_SetString(PyExc_OverflowError,
                        "code: locals + stack size is too large");
        return -1;
    }
```

What do you think?

**Comment by picnixz** [comment]
Ok, with this new limit, I don't trigger the crash on free-threaded builds (it's just that the code (with another evil value) takes forever to complete but that's expected). I've removed the tests because they won't work anymore (it's not possible to bypass the check in code objects). 

However, for safety, I kept the checks when computing `__sizeof__` if `Py_SSIZE_T_MAX < INT_MAX` which may happen on some platforms (POSIX does not specify any restriction or relation between `ssize_t` and `int` ranges) and the upper limit may be too crude (it's not really possible to refine it when constructing code objects for frames).

**Comment by picnixz** [comment]
Ok, with this new limit, I don't trigger the crash on free-threaded builds (it's just that the code (with another evil value) takes forever to complete but that's expected). I've removed the tests because they won't work anymore (it's not possible to bypass the check in code objects). 

However, for safety, I kept the checks when computing `__sizeof__` if `Py_SSIZE_T_MAX < INT_MAX` which may happen on some platforms (POSIX does not specify any restriction or relation between `ssize_t` and `int` ranges) and the upper limit may be too crude (it's not really possible to refine it when constructing code objects for frames).

I have made the requested changes; please review again.

File: Objects/codeobject.c [hunk 7]
```diff
@@ -443,12 +443,14 @@ _PyCode_Validate(struct _PyCodeConstructor *con)
      * usually prevents crashes due to assertions but a MemoryError may still
      * be triggered later.
      *
-     * See https://github.com/python/cpython/issues/126119 for details.
+     * See https://github.com/python/cpython/issues/126119 for details
+     * and corresponding PR for the rationale on the upper limit value.
      */
-    int ub = (int)(INT_MAX / sizeof(PyObject *)) - FRAME_SPECIALS_SIZE;
+    Py_ssize_t limit = (Py_ssize_t)(INT_MAX / 16) - FRAME_SPECIALS_SIZE;
```
Comment (markshannon) [inline Objects/codeobject.c]: No need for the `FRAME_SPECIALS_SIZE`. That's why it is `INT_MAX / 16` not `INT_MAX/8`, to make sure there is *plenty* of headroom.

File: Objects/frameobject.c [hunk 8]
```diff
@@ -1802,11 +1802,25 @@ PyDoc_STRVAR(clear__doc__,
 static PyObject *
 frame_sizeof(PyFrameObject *f, PyObject *Py_UNUSED(ignored))
 {
-    Py_ssize_t res;
-    res = offsetof(PyFrameObject, _f_frame_data) + offsetof(_PyInterpreterFrame, localsplus);
+    Py_ssize_t base = offsetof(PyFrameObject, _f_frame_data)
```
Comment (markshannon) [inline Objects/frameobject.c]: I think there is no need to change `frame_sizeof` as `_PyCode_Validate` ensures we can't overflow.

File: Objects/genobject.c [hunk 9]
```diff
@@ -817,11 +817,25 @@ static PyMemberDef gen_memberlist[] = {
 static PyObject *
 gen_sizeof(PyGenObject *gen, PyObject *Py_UNUSED(ignored))
 {
-    Py_ssize_t res;
-    res = offsetof(PyGenObject, gi_iframe) + offsetof(_PyInterpreterFrame, localsplus);
+    Py_ssize_t base = offsetof(PyGenObject, gi_iframe)
```
Comment (markshannon) [inline Objects/genobject.c]: Likewise, `_PyCode_Validate()` prevents overflow here.

File: Objects/codeobject.c [hunk 10]
```diff
@@ -436,7 +436,23 @@ _PyCode_Validate(struct _PyCodeConstructor *con)
         PyErr_SetString(PyExc_ValueError, "code: co_varnames is too small");
         return -1;
     }
-
+    /*
+     * The framesize = stacksize + nlocalsplus + FRAME_SPECIALS_SIZE is used
```
Comment (markshannon) [inline Objects/codeobject.c]: This comment needs updating.

**Review by markshannon — CHANGES_REQUESTED** [review]
We should be able to remove the changes to `frame_sizeof` and `gen_sizeof` as `_PyCode_Validate()` protects from overflow.

**Comment by picnixz** [comment]
I've updated the comment and removed the un-necessary checks. Please tell me if the comment is still imprecise or if you have a preferred formulation.

I have made the requested changes; please review again.

**Comment by picnixz** [comment]
Friendly ping @markshannon (I don't really want to merge this without your explicit approval)

**Comment by gaogaotiantian** [comment]
Hmm, I still think Mark should be the one that approves this PR before it's merged. I don't see obvious issues for the code change. However, is the `INT_MAX / 16` a precise limit? Both our checks and tests are calculating very precisely for this number, the test even does +1/-1 level validation. Not saying it's bad, but is it necessary? We want to prevent an artificial code object that has `INT_MAX` level locals?

Again, I'm not opposing the current code, that's just the first thought I saw it.

**Comment by picnixz** [comment]
> Both our checks and tests are calculating very precisely for this number, the test even does +1/-1 level validation

The +1/-1 validation is simply to check that we didn't change the formula. We can take any bound we want instead of just INT_MAX / 16, as soon as it's a valid one. But the bound is not tight. Mark said that we could have a tighter bound (so no, it's not a precise limit, it can be a bit bigger in general), namely

> We could increase the limit to something like $INT\_MAX/8 - K$ where $K$ is roughly the base frame size, but I don't see any advantage to doing so and it would be more fragile.

> want to prevent an artificial code object that has INT_MAX level locals?

We just want to prevent a value of `co_stacksize` being too large because we're doing some arithmetic with ints behind and use it in a loop. 

cc @markshannon

File: Lib/test/test_code.py [hunk 11]
```diff
@@ -1115,6 +1121,25 @@ def test_stateless(self):
...
+
+        c = foo.__code__
+
+        co_nlocalsplus = len({*c.co_varnames, *c.co_cellvars, *c.co_freevars})
+        # anything below that limit is a valid co_stacksize
+        evil_stacksize = int(_testcapi.INT_MAX / 16 - co_nlocalsplus)
+
+        with self.assertRaisesRegex(OverflowError, "stack size is too large"):
+            c.__replace__(co_stacksize=evil_stacksize)
+        c.__replace__(co_stacksize=evil_stacksize - 1)
```
Comment (markshannon) [inline Lib/test/test_code.py]: We shouldn't be hardcoding the exact limit here.
Maybe `too_large_stacksize = _testcapi.INT_MAX // 16` and `ok_stacksize = too_large_stacksize // 2`

File: Misc/NEWS.d/next/Core_and_Builtins/2024-10-29-11-47-19.gh-issue-126119.xbAvxt.rst [hunk 12]
```diff
@@ -0,0 +1,4 @@
+Fix a crash in DEBUG builds due to an overflow when the :attr:`co_stacksize
```
Comment (markshannon) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-10-29-11-47-19.gh-issue-126119.xbAvxt.rst]: I'd rephrase this as fixing a lack of overflow checking in `code.replace` since the bytecode compiler will never create a code object like this.

**Review by markshannon — CHANGES_REQUESTED** [review]
The code looks good. 
A couple of quibbles about the test and news entry, otherwise good to merge.

**Comment by markshannon** [comment]
Sorry for the delay in the review, and thanks for doing this.

**Comment by picnixz** [comment]
I have made the requested changes; please review again.

**Comment by picnixz** [comment]
@markshannon Can you have a final look at this one?

**Comment by picnixz** [comment]
Considering this hasn't moved and considering we now don't really consider crashes messing up with the `__replace__` to be valid concerns, I'm going to close this one.
